SVD pipeline for DMD-ERA5

dsj976 opened the pull request:
This pull request introduces significant enhancements and new features to the `dmd_era5` project, primarily focusing on the integration of Singular Value Decomposition (SVD) operations on ERA5 datasets. The main changes include updates to configuration handling, the addition of new dependencies, and the implementation of new functions for SVD processing.

### Configuration and Dependency Updates:
* Added `scikit-learn` and `dask` to the dependencies in `pyproject.toml` and `.pre-commit-config.yaml`. [[1]](diffhunk://#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R46) [[2]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R37-R38)
* Introduced a new configuration section `[era5-svd]` in `config.ini` with parameters for SVD processing.

### New Functions and Modules:
* Added new functions `apply_delay_embedding`, `flatten_era5_variables`, and `space_coord_to_level_lat_lon` to the `slice_tools` module. [[1]](diffhunk://#diff-6a74c7a9e542fcaa49f56960f371edfb135e7d76a33ef67ffbdcc4fde4b6b445L10-R24) [[2]](diffhunk://#diff-bae6e6d1f6a6b89e03ca4626031b92d46ed05a7c077231d099f7c4e9de65ec75R2-R18)
* Implemented `create_mock_era5_svd` function in the `create_mock_data` module for generating mock SVD datasets.
* Added `svd_on_era5`, `combine_svd_results`, `retrieve_era5_slice`, `retrieve_svd_results`, `add_config_attributes`, and `main` functions to the `era5_svd` module.

### Enhancements to Existing Functions:
* Updated `config_parser` to support the new `era5-svd` section and validate its specific fields. [[1]](diffhunk://#diff-5adb5891e8af8cd4db4e54314f9dd9b429bbf5245a3b8fcbacd2bd7ffc876f10R58-R71) [[2]](diffhunk://#diff-5adb5891e8af8cd4db4e54314f9dd9b429bbf5245a3b8fcbacd2bd7ffc876f10L73-R95) [[3]](diffhunk://#diff-5adb5891e8af8cd4db4e54314f9dd9b429bbf5245a3b8fcbacd2bd7ffc876f10L172-R288)
* Modified `retrieve_data_from_dvc` to handle the retrieval of SVD results in addition to ERA5 slices. [[1]](diffhunk://#diff-04701341533595143a9306fe4ee44814f4291aee9e7780d50a86d95beddd12f1L133-R165) [[2]](diffhunk://#diff-04701341533595143a9306fe4ee44814f4291aee9e7780d50a86d95beddd12f1L150-R178) [[3]](diffhunk://#diff-04701341533595143a9306fe4ee44814f4291aee9e7780d50a86d95beddd12f1L162-R207) [[4]](diffhunk://#diff-04701341533595143a9306fe4ee44814f4291aee9e7780d50a86d95beddd12f1L180-R235)

dsj976 commented:
A bit of a long PR, but would be grateful if you could give it a quick look and test that it works for you on Baskerville @ots22 @louisavz.

ots22 commented:
Thanks @dsj976 - I'll be able to have a go tomorrow or maybe later today

dsj976 commented:
Hi @ots22 @louisavz - would be great if you could have a quick look by the end of the week! Thanks

ots22 commented:
Sorry David, I've just tried on Baskerville. All works as expected following the README.

A note on the output files would be helpful - I can see data/era5_svd is created, but maybe a comment on what it contains? It could just be the output of loading it with xarray and pointing out U, s, V etc.

louisavz commented on README.md:
Line 15 should be `cd DMD-ERA5`

louisavz commented on README.md:
The step `make dvc-init` might be redundant, I ran into an error saying 

```(.venv) $ make dvc-init
dvc init
ERROR: failed to initiate DVC - '.dvc' exists. Use `-f` to force.
make: *** [Makefile:25: dvc-init] Error 1

dsj976 replied:
I mean that you only need to call `make dvc-init` once in a given repository. This is essentially the same as calling `git init` to initialize a Git repo. If the `.git` folder already exists, it will complain if you run `git init` again.

louisavz replied:
I only called `make dvc-init` once. To avoid having residual memory hold-over, I created a completely new directory for testing. I also checked to ensure there wasn't a .dvc file before starting. However, Oliver didn't run into this problem, I think it should be okay. If I remember correctly, the .dvc file was created during a prior step (`make dvc-setup`)

dsj976 replied:
Oh I see! Yes, you either do one or the other. I have clarified that now in the README. `make dvc-setup` does the whole set up, including setting up a remote DVC repo. If you don't want to set up a remote repo, you do the steps one by one, the first of which is calling `make dvc-init`.

louisavz replied:
Ah! Got it, I misunderstood. Thanks for clarifying!

louisavz commented on README.md:
````diff
@@ -78,6 +78,22 @@ Once you have pushed to the remote, it is safe to delete the downloaded NetCDF f
+This package makes use of the optimized DMD algorithm proposed by Askham and Kutz (2018). An optional step of the algorithm that is suitable for large datasets is to perform Singular Value Decomposition (SVD) on the data prior to running DMD. `dmd-era5` allows you to perform SVD on ERA5 data by running:
+
+```bash
+python -m dmd_era5.era5_svd.era5_svd
````
Prior to this step on Baskerville, I had to set up path to src (stopping at DMD-ERA5 was not enough). This might be a packaging issue with dmd-era5. Perhaps consider adding `packages = [{include = "dmd_era5", from = "src"}]` to pyproject.toml.

`export PYTHONPATH=$PYTHONPATH:/bask/homes/s/sgft6713/tmp/DMD-ERA5/src/`

dsj976 replied:
Hmmm... I think this happened because you were not using a Python virtual environment? Looks like a temporary folder was created in your home directory when you tried to install the package. I have tried reinstalling the package from scratch on Baskerville (inside a virtual environment) and it has worked fine for me. I have added the virtual environment creation step to the README.

louisavz replied:
Mmmm... I was using a virtual environment the second time around (I didn't get this far without a .venv). The "tmp" folder was me making a new folder to test the code from scratch, rather than pulling the remote branch, on purpose.

louisavz commented:
@dsj976 Looks really good to me! Apologies for not seeing this PR needing review until yesterday. I've tried running this on Baskerville and was able to run to completion after making the above changes. Please can you take a look at the above comments?

dsj976 commented:
Thanks @ots22 and @louisavz for your review. I have addressed comments and added description of the contents of the SVD NetCDF file to the README (see commits https://github.com/ClimeTrend/DMD-ERA5/pull/27/commits/eef205f2d051d1807a13add8a169e190218af70f and https://github.com/ClimeTrend/DMD-ERA5/pull/27/commits/f71d29c79265882aba438cdc6952256bd626b40b). If you could have a quick look at this and approve so we can merge this. Thanks!

ots22 commented on media/svd_netcdf_contents.png:
A text version of the output would be preferable to text in an image (accessibility, easier copy/paste).

dsj976 replied:
Agreed. Will tackle this in a future PR.
